File: scripts/tech-ds/distribution/phase-space-bridge.js
// Copyright (C) 2020 abomb4
//
// This file is part of Dimension Shard.
//
// Dimension Shard is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Dimension Shard is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Dimension Shard.  If not, see <http://www.gnu.org/licenses/>.

const lib = require('abomb4/lib');
const items = require('ds-common/items');

const lastBuildInvalidate = 60 * 10;
let globalLastBuildTime = 0;
let phaseSpaceBridge = extend(ItemBridge, 'phase-space-bridge', {

    drawPlace(x, y, rotation, valid) {
        const range = this.range;
        const tilesize = Vars.tilesize;
        Drawf.dashCircle(x * tilesize, y * tilesize, range * tilesize, Pal.accent);

        // check if a mass driver is selected while placing this driver
        if (!Vars.control.input.frag.config.isShown()) return;
        let selected = Vars.control.input.frag.config.getSelectedTile();
        if (selected == null || (selected.block != this) || !(selected.within(x * tilesize, y * tilesize, range * tilesize))) return;

        // if so, draw a dotted line towards it while it is in range
        let sin = Mathf.absin(Time.time, 6, 1);
        Tmp.v1.set(x * tilesize + this.offset, y * tilesize + this.offset).sub(selected.x, selected.y).limit((this.size / 2 + 1) * tilesize + sin + 0.5);
        let x2 = x * tilesize - Tmp.v1.x, y2 = y * tilesize - Tmp.v1.y,
            x1 = selected.x + Tmp.v1.x, y1 = selected.y + Tmp.v1.y;
        let segs = (selected.dst(x * tilesize, y * tilesize) / tilesize);

        Lines.stroke(2, Pal.gray);
        Lines.dashLine(x1, y1, x2, y2, segs);
        Lines.stroke(1, Pal.placing);
        Lines.dashLine(x1, y1, x2, y2, segs);
        Draw.reset();
    },
    linkValid(tile, other, checkDouble) {
        if (other == null || tile == null) return false;

        if (checkDouble === undefined) { checkDouble = true; }
        return other.block() === phaseSpaceBridge
            && tile.block() === phaseSpaceBridge
            && (!checkDouble || other.build.link != tile.pos())
            && this.positionsValid(tile.x, tile.y, other.x, other.y);
    },
    positionsValid(x1, y1, x2, y2) {
        return Mathf.dst(x1, y1, x2, y2) <= this.range;
    },
    changePlacementPath(points, rotation) {
        Placement.calculateNodes(points, this, rotation, lib.boolf2((point, other) => this.positionsValid(point.x, point.y, other.x, other.y)));
    },
    findLink(x, y) {
        if (Time.time - globalLastBuildTime > lastBuildInvalidate) {
            return null;
        }
        return this.super$findLink(x, y);
    },
});
phaseSpaceBridge.buildVisibility = BuildVisibility.shown;
phaseSpaceBridge.category = Category.distribution;
phaseSpaceBridge.size = 1;
phaseSpaceBridge.health = 220;
phaseSpaceBridge.hasItems = true;
phaseSpaceBridge.hasLiquids = true;
phaseSpaceBridge.outputsLiquid = true;
phaseSpaceBridge.itemCapacity = 25;
phaseSpaceBridge.liquidCapacity = 20;
phaseSpaceBridge.liquidPressure = 1.2;
phaseSpaceBridge.range = 15;
phaseSpaceBridge.transportTime = 0.01;
phaseSpaceBridge.requirements = ItemStack.with(
    Items.metaglass, 15,
    Items.silicon, 5,
    Items.titanium, 8,
    Items.phaseFabric, 15,
    items.dimensionShard, 20
);
phaseSpaceBridge.consumes.power(0.5);

Events.on(BlockBuildBeginEvent, cons(e => {
    if (Vars.player != null && Vars.player.unit() == e.unit && e.tile.block() != phaseSpaceBridge) {
        phaseSpaceBridge.lastBuild = null;
    }
}));

lib.setBuildingSimple(phaseSpaceBridge, ItemBridge.ItemBridgeBuild, block => ({

    playerPlaced(config) {
        this.super$playerPlaced(config);
        globalLastBuildTime = Time.time;
    },
    drawConfigure() {
        let entity = this;
        let tile = this.tile;
        let block = this.block;
        let x = this.x;
        let y = this.y;

        Draw.color(Pal.accent);
        Lines.stroke(1);
        Lines.square(x, y, block.size * Vars.tilesize / 2 + 1);

        let target;
        if (entity.link != -1 && (target = Vars.world.tile(entity.link)) != null && this.block.linkValid(tile, target, true)) {
            Draw.color(Pal.place);
            Lines.square(target.x * Vars.tilesize, target.y * Vars.tilesize, block.size * Vars.tilesize / 2 + 1 + (Mathf.absin(Time.time, 4, 1)));
        }

        Drawf.dashCircle(x, y, this.block.range * Vars.tilesize, Pal.accent);
    },
    draw() {
        Draw.rect(this.block.region, this.x, this.y, this.block.rotate ? this.rotdeg() : 0);
        this.drawTeamTop();

        Draw.z(Layer.power);
        // Link each
        const tile = this.tile;
        const tilesize = Vars.tilesize;
        let entity = this;

        let other = Vars.world.tile(entity.link);
        if (!this.block.linkValid(tile, other)) return;
        let otherBuild = other.build;
        if (otherBuild == null) { return; }

        let opacity = Core.settings.getInt("bridgeopacity") / 100;
        if (Mathf.zero(opacity)) return;

        // draw it

        let angle = Angles.angle(tile.worldx(), tile.worldy(), other.worldx(), other.worldy());
        Draw.color(Color.white, Color.black, Mathf.absin(Time.time, 6, 0.07));
        Draw.alpha(Math.max(entity.uptime, 0.25) * opacity);

        Draw.rect(this.block.endRegion, this.x, this.y, angle + 90);
        Draw.rect(this.block.endRegion, otherBuild.x, otherBuild.y, angle + 270);

        Lines.stroke(8);
        Lines.line(this.block.bridgeRegion, tile.worldx(), tile.worldy(), other.worldx(), other.worldy(), false);

        let dist = Math.max(Math.abs(other.x - tile.x), Math.abs(other.y - tile.y));

        let time = entity.time2 / 1.7;
        let arrows = (dist) * tilesize / 4 - 2;

        Draw.color();

        for (let a = 0; a < arrows; a++) {
            Draw.alpha(Mathf.absin(a / arrows - entity.time / 100, 0.1, 1) * entity.uptime * opacity);
            Draw.rect(this.block.arrowRegion,
                tile.worldx() + Angles.trnsx(angle, (tilesize / 2 + a * 4 + time % 4)),
                tile.worldy() + Angles.trnsy(angle, (tilesize / 2 + a * 4 + time % 4)), angle);
        }
        Draw.reset();
    },
    canDump(to, item) {
        // 4 direction output
        return true;
    },
    acceptItem(source, item) {
        const tile = this.tile;
        if (this.team != source.team) return false;
        const itemCapacity = this.block.itemCapacity;

        let entity = this;
        let other = Vars.world.tile(entity.link);

        if (this.block.linkValid(tile, other)) {
            return this.items.total() < itemCapacity;
        } else {
            return source.block == this.block && source.link == tile.pos() && this.items.total() < itemCapacity;
        }

    },
    updateTile() {
        this.super$updateTile();
        // Try dump liquid if not be connected
        let entity = this;
        let tile = this.tile;
        let other = Vars.world.tile(entity.link);
        if (!this.block.linkValid(tile, other)) {
            this.dumpLiquid(entity.liquids.current());
        }
    },
    acceptLiquid(source, liquid) {
        if (this.team != source.team || !this.block.hasLiquids) return false;

        let entity = this;
        let tile = this.tile;
        let other = Vars.world.tile(entity.link);

        if (this.block.linkValid(tile, other)) {
            return true;
        } else if (!(source.block == this.block && source.link == tile.pos())) {
            return false;
        }

        return this.liquids.get(liquid) < this.block.liquidCapacity
            && (this.liquids.current() == liquid || this.liquids.get(this.liquids.current()) < 0.2);
    },
    updateTransport(other) {
        let entity = this;

        if (entity.uptime >= 0.5 && entity.timer.get(this.block.timerTransport, this.block.transportTime)) {
            // transport items
            let item = entity.items.take();
            if (item != null && other.acceptItem(this, item)) {
                other.handleItem(this, item);
                entity.cycleSpeed = Mathf.lerpDelta(entity.cycleSpeed, 4, 0.05);
            } else {
                entity.cycleSpeed = Mathf.lerpDelta(entity.cycleSpeed, 1, 0.01);
                if (item != null) entity.items.add(item, 1);
            }

            // transport liquid
            this.moveLiquid(other, this.liquids.current());
        }
    },
    checkDump(to){
        let other = Vars.world.tile(this.link);
        return (!this.block.linkValid(this.tile, other, false));
    }
}));

exports.phaseSpaceBridge = phaseSpaceBridge;

File: main/src/dimensionshard/types/blocks/PhaseSpaceBridge.java
package dimensionshard.types.blocks;

import arc.Events;
import arc.graphics.Color;
import arc.graphics.g2d.Draw;
import arc.graphics.g2d.Lines;
import arc.math.Angles;
import arc.math.Mathf;
import arc.math.geom.Point2;
import arc.struct.Seq;
import arc.util.Time;
import arc.util.Tmp;
import mindustry.Vars;
import mindustry.core.Renderer;
import mindustry.game.EventType;
import mindustry.gen.Building;
import mindustry.graphics.Drawf;
import mindustry.graphics.Layer;
import mindustry.graphics.Pal;
import mindustry.input.Placement;
import mindustry.type.Item;
import mindustry.world.Tile;
import mindustry.world.blocks.distribution.ItemBridge;

import static mindustry.Vars.tilesize;
import static mindustry.Vars.world;

/**
 * 相位空间桥，多方向传输物品
 *
 * @author abomb4 2022-10-07
 */
public class PhaseSpaceBridge extends ItemBridge {

    /** 超过 10s 后，再次建造时不会自动连接 */
    public final float lastBuildInvalidate = 60 * 10;
    /** 上次建造时间 */
    public float globalLastBuildTime = 0;

    public PhaseSpaceBridge(String name) {
        super(name);
    }

    @Override
    public void drawPlace(int x, int y, int rotation, boolean valid) {
        // if (!dsGlobal.techDsAvailable()) {
        //     this.drawPlaceText(lib.getMessage("msg", "dimensionCoreRequired"), x, y, valid);
        // }
        var range = this.range;
        var tilesize = Vars.tilesize;
        Drawf.dashCircle(x * tilesize, y * tilesize, range * tilesize, Pal.accent);

        // check if a mass driver is selected while placing this driver
        if (!Vars.control.input.config.isShown()) {
            return;
        }
        var selected = Vars.control.input.config.getSelected();
        if (selected == null || (selected.block != this) ||
            !(selected.within(x * tilesize, y * tilesize, range * tilesize))) {
            return;
        }

        // if so, draw a dotted line towards it while it is in range
        float sin = Mathf.absin(Time.time, 6, 1);
        Tmp.v1.set(x * tilesize + this.offset, y * tilesize + this.offset)
            .sub(selected.x, selected.y)
            .limit((this.size / 2F + 1) * tilesize + sin + 0.5F);
        float x2 = x * tilesize - Tmp.v1.x, y2 = y * tilesize - Tmp.v1.y,
            x1 = selected.x + Tmp.v1.x, y1 = selected.y + Tmp.v1.y;
        int segs = (int) (selected.dst(x * tilesize, y * tilesize) / tilesize);

        Lines.stroke(2, Pal.gray);
        Lines.dashLine(x1, y1, x2, y2, segs);
        Lines.stroke(1, Pal.placing);
        Lines.dashLine(x1, y1, x2, y2, segs);
        Draw.reset();
    }

    @Override
    public boolean linkValid(Tile tile, Tile other, boolean checkDouble) {
        if (other == null || tile == null || tile == other || !this.positionsValid(tile.x, tile.y, other.x, other.y)) {
            return false;
        }

        return other.block() == this
            && tile.block() == this
            && (!checkDouble || ((PhaseSpaceBridgeBuild) other.build).link != tile.pos());
    }

    @Override
    public boolean positionsValid(int x1, int y1, int x2, int y2) {
        return Mathf.dst(x1, y1, x2, y2) <= this.range;
    }

    @Override
    public void changePlacementPath(Seq<Point2> points, int rotation) {
        Placement.calculateNodes(points, this, rotation,
            (point, other) -> this.positionsValid(point.x, point.y, other.x, other.y));
    }

    @Override
    public Tile findLink(int x, int y) {
        // 自动链接功能，超过指定时间后再次建造不会自动链接
        if (Time.time - this.globalLastBuildTime > this.lastBuildInvalidate) {
            return null;
        }
        return super.findLink(x, y);
    }

    @Override
    public void load() {
        super.load();

        Events.on(EventType.BlockBuildBeginEvent.class, (e -> {
            if (Vars.player != null && Vars.player.unit() == e.unit && e.tile.block() != this) {
                this.lastBuild = null;
            }
        }));
    }

    /** Instance */
    public class PhaseSpaceBridgeBuild extends ItemBridge.ItemBridgeBuild {

        @Override
        public void playerPlaced(Object config) {
            super.playerPlaced(config);
            PhaseSpaceBridge.this.globalLastBuildTime = Time.time;
        }

        @Override
        public void drawConfigure() {
            Tile tile = this.tile;
            float x = this.x;
            float y = this.y;

            Draw.color(Pal.accent);
            Lines.stroke(1);
            PhaseSpaceBridge block = (PhaseSpaceBridge) this.block;
            Lines.square(x, y, block.size * Vars.tilesize / 2.0F + 1);

            Tile target;
            if (this.link != -1
                && (target = Vars.world.tile(this.link)) != null
                && block.linkValid(tile, target, true)) {
                Draw.color(Pal.place);
                Lines.square(
                    target.x * Vars.tilesize,
                    target.y * Vars.tilesize,
                    block.size * Vars.tilesize / 2.0F + 1 + (Mathf.absin(Time.time, 4, 1)));
            }

            Drawf.dashCircle(x, y, block.range * Vars.tilesize, Pal.accent);
        }

        @Override
        public void draw() {
            Draw.rect(this.block.region, this.x, this.y, this.block.rotate ? this.rotdeg() : 0);
            this.drawTeamTop();
            Draw.z(Layer.power);

            Tile other = world.tile(link);
            if (!linkValid(tile, other)) {
                return;
            }

            if (Mathf.zero(Renderer.bridgeOpacity)) {
                return;
            }

            int i = relativeTo(other.x, other.y);

            if (pulse) {
                Draw.color(Color.white, Color.black, Mathf.absin(Time.time, 6f, 0.07f));
            }

            float warmup = hasPower ? this.warmup : 1f;

            Draw.alpha((fadeIn ? Math.max(warmup, 0.25f) : 1f) * Renderer.bridgeOpacity);

            var angle = Angles.angle(tile.worldx(), tile.worldy(), other.worldx(), other.worldy());

            Draw.rect(endRegion, x, y, angle + 90);
            Draw.rect(endRegion, other.drawx(), other.drawy(), angle + 270);

            Lines.stroke(bridgeWidth);
            Lines.line(bridgeRegion, tile.worldx(), tile.worldy(), other.worldx(), other.worldy(), false);

            int dist = Math.max(Math.abs(other.x - tile.x), Math.abs(other.y - tile.y)) - 1;

            Draw.color();

            int arrows = dist * 2 - 2;

            for (int a = 0; a < arrows; a++) {
                Draw.alpha(Mathf.absin(a - time / arrowTimeScl, arrowPeriod, 1f) * warmup * Renderer.bridgeOpacity);
                Draw.rect(arrowRegion,
                    tile.worldx() + Angles.trnsx(angle, (tilesize / 2f + a * 4 + time % 4)),
                    tile.worldy() + Angles.trnsy(angle, (tilesize / 2f + a * 4 + time % 4)), angle);
            }

            Draw.reset();
        }

        @Override
        public boolean canDump(Building to, Item item) {
            // 4 directions output
            return true;
        }

        @Override
        public boolean acceptItem(Building source, Item item) {
            if (this.team != source.team) {
                return false;
            }

            Tile other = Vars.world.tile(this.link);
            PhaseSpaceBridge block = (PhaseSpaceBridge) this.block;

            if (block.linkValid(this.tile, other)) {
                return this.items.total() < PhaseSpaceBridge.this.itemCapacity;
            } else {
                return source.block == this.block
                    && ((PhaseSpaceBridgeBuild) source).link == this.tile.pos()
                    && this.items.total() < PhaseSpaceBridge.this.itemCapacity;
            }
        }

        @Override
        public void updateTile() {
            super.updateTile();
            // Try dump liquid if not be connected
            Tile other = Vars.world.tile(this.link);
            PhaseSpaceBridge block = (PhaseSpaceBridge) this.block;

            // 如果没去连别人，则 dump liquid
            if (!block.linkValid(this.tile, other)) {
                this.dumpLiquid(this.liquids.current());
            }
        }

        @Override
        protected boolean checkAccept(Building source, Tile other) {
            if (this.tile == null || this.linked(source)) {
                return true;
            }
            return PhaseSpaceBridge.this.linkValid(this.tile, other) || other == this.tile;
        }

        @Override
        public void updateTransport(Building other) {

            this.transportCounter += this.edelta();
            while (this.transportCounter >= PhaseSpaceBridge.this.transportTime) {
                Item item = this.items.take();
                if (item != null && other.acceptItem(this, item)) {
                    other.handleItem(this, item);
                    this.moved = true;
                } else if (item != null) {
                    this.items.add(item, 1);
                    this.items.undoFlow(item);
                }
                this.transportCounter -= PhaseSpaceBridge.this.transportTime;

                // transport liquid
                this.moveLiquid(other, this.liquids.current());
            }
        }

        @Override
        protected boolean checkDump(Building to) {
            Tile other = Vars.world.tile(this.link);
            return (!PhaseSpaceBridge.this.linkValid(this.tile, other, false));
        }
    }
}
